File: C++_Version/Tree/257Binary_Tree_Paths.cpp
//Given a binary tree, return all root - to - leaf paths.
//
//For example, given the following binary tree :
//
//1
/// \
//2     3
//\
//5
//All root - to - leaf paths are :
//
//["1->2->5", "1->3"]

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
#include <string>
#include <unordered_map>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<string> binaryTreePaths(TreeNode* root) {
		if (root == NULL)
			return vector<string>();
		
		vector<string> result;

		binaryTreePathsHelper(root, result, std::to_string(root->val) );

		return result;
	}

private:
	void binaryTreePathsHelper(TreeNode* node, vector<string>& result, string formal_string)
	{
		if (node->left == NULL && node->right == NULL)
		{
			result.push_back(formal_string);
			return;
		}

		if (node->left != NULL)
			binaryTreePathsHelper(node->left, result, formal_string + "->" + std::to_string(node->left->val));
		if (node->right != NULL)
			binaryTreePathsHelper(node->right, result, formal_string + "->" + std::to_string(node->right->val));
	}
};


int main()
{
	TreeNode* root = new TreeNode(1);
	TreeNode* node1 = new TreeNode(2);
	TreeNode* node2 = new TreeNode(3);
	TreeNode* node3 = new TreeNode(5);
	TreeNode* node4 = new TreeNode(7);
	root->left = node1;
	root->right = node2;
	node1->right = node3;

	Solution sol1;

	vector<string> receive = sol1.binaryTreePaths(root);
	
	system("pause");

	return 0;
}

File: Java_Version/257_Binary_Tree_Paths.java
/*
  Given a binary tree, return all root-to-leaf paths.

  For example, given the following binary tree:

            1
          /   \
         2     3
          \
           5

  All root-to-leaf paths are:

  ["1->2->5", "1->3"]
*/

import java.util.*;
import java.lang.Math;
import java.lang.System;
import java.lang.Integer;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}


class Solution {

    public static void main(String[] args) {

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.right = new TreeNode(5);

        List<String> result = new Solution().binaryTreePaths(root);

        System.out.println();
    }

    public List<String> binaryTreePaths(TreeNode root) {

        List<String> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        binaryTreePathsHelper(root, new ArrayList<>(), result);
        return result;
    }

    private void binaryTreePathsHelper(
            TreeNode root, List<Integer> existingNumbers, List<String> result) {

        if (root == null) {
            return;
        }

        existingNumbers.add(root.val);

        if (root.left == null && root.right == null) {
            result.add(convertIntegerListToString(existingNumbers));
        }

        binaryTreePathsHelper(root.left, existingNumbers, result);
        binaryTreePathsHelper(root.right, existingNumbers, result);

        existingNumbers.remove(existingNumbers.size() - 1);
    }

    private String convertIntegerListToString(List<Integer> numbers) {

        StringBuilder result = new StringBuilder();

        for (int i = 0; i < numbers.size() - 1; i++) {

            result.append(numbers.get(i));
            result.append("->");
        }

        result.append(numbers.get(numbers.size() - 1));

        return result.toString();
    }
}
